python/shotgun_model/shotgun_entity_model.py

```diff
@@ -0,0 +1,258 @@
+                        from the specified item.
+        """
+        entities = []
+        current_item = item
+
+        # first, if this is a leaf item then it will represent an entity:
+        sg_data = current_item.get_sg_data()
+        if sg_data:
+            entities.append(sg_data)
+            current_item = current_item.parent()
+
+        # now walk up the tree and look for an entity in the fields of the 
+        # parent items:
+        while current_item:
+            field_data = get_sanitized_data(current_item, self.SG_ASSOCIATED_FIELD_ROLE)
+            field_value = field_data.get("value")
+            if (field_value 
+                and isinstance(field_value, dict) 
+                and "id" in field_value 
+                and "type" in field_value):
+                entities.append(field_value)
+            current_item = current_item.parent()
+
+        return entities
+
+    def get_entity(self, item):
+        """
+        Get the Shotgun entity details for the specified model item.
+
+        :param item:    The item to retrieve the entity details for
+        :returns:       A Shotgun entity dictionary for the item if it represents
+                        and entity, otherwise None
+        """
+        sg_data = item.get_sg_data()
+        if sg_data:
+            return sg_data
+
+        field_data = get_sanitized_data(item, self.SG_ASSOCIATED_FIELD_ROLE)
+        field_value = field_data.get("value")
+        if (field_value 
+            and isinstance(field_value, dict) 
+            and "id" in field_value 
+            and "type" in field_value):
+            return field_value
+
+        return None
+
+    def async_refresh(self):
+        """
+        Trigger an asynchronous refresh of the model
+        """
+        self._refresh_data()
+
+    def _populate_default_thumbnail(self, item):
+        """
+        Whenever an item is constructed, this methods is called. It allows subclasses to intercept
+        the construction of a QStandardItem and add additional metadata or make other changes
+        that may be useful. Nothing needs to be returned.
+
+        :param item: QStandardItem that is about to be added to the model. This has been primed
+                     with the standard settings that the ShotgunModel handles.
+        :param sg_data: Shotgun data dictionary that was received from Shotgun given the fields
+                        and other settings specified in load_data()
+        """
+        found_icon = False
+
+        # get the associated field data with this node
+        field_data = get_sanitized_data(item, self.SG_ASSOCIATED_FIELD_ROLE)
+        # get the full sg data for this node (leafs only)
+        sg_data = get_sg_data(item)
+
+        # {'name': 'sg_sequence', 'value': {'type': 'Sequence', 'id': 11, 'name': 'bunny_080'}}
+        field_value = field_data["value"]
+
+        if isinstance(field_value, dict) and "name" in field_value and "type" in field_value:
```

Comment:
the code needed isn't that bad and tbh there are so many variations that you'd end up having to read up on the predicate every time you came across it to remind yourself what it checks...